pkg/kubernetes/deployment/v1alpha1/kubernetes.go

```diff
@@ -0,0 +1,199 @@
+//  of deployment instances of AppsV1 api group
+type rollOutStatusAppsV1Fn func(cli *kubernetes.Clientset, name, namespace string) ([]byte, error)
+
+// kubeclient enables kubernetes API operations on deployment instance
+type kubeclient struct {
+	// clientset refers to kubernetes clientset. It is responsible to
+	// make kubernetes API calls for crud op
+	clientset *kubernetes.Clientset
+
+	// functions useful during mocking
+	getClientset getClientsetFn
+
+	getExtnV1Beta1 getExtnV1Beta1Fn
+
+	getAppsV1 getAppsV1Fn
+
+	rollOutStatusExtnV1Beta1 rollOutStatusExtnV1Beta1Fn
+
+	rollOutStatusAppsV1 rollOutStatusAppsV1Fn
+}
+
+// rolloutOutput struct contaons message and boolean value to show rolloutstatus
+type rolloutOutput struct {
+	IsRolledout bool   `json:"IsRolledout"`
+	Message     string `json:"Message"`
+}
+
+// kubeclientBuildOption defines the abstraction to build a kubeclient instance
+type kubeclientBuildOption func(*kubeclient)
+
+// withDefaults sets the default options of kubeclient instance
+func (k *kubeclient) withDefaults() {
+	if k.getClientset == nil {
+		k.getClientset = func() (clients *kubernetes.Clientset, err error) {
+			config, err := kclient.Config().Get()
+			if err != nil {
+				return nil, err
+			}
+			return kubernetes.NewForConfig(config)
+		}
+	}
+
+	if k.getExtnV1Beta1 == nil {
+		k.getExtnV1Beta1 = func(cli *kubernetes.Clientset, name,
+			namespace string) (d *api_extn_v1beta1.Deployment, err error) {
+			d = &api_extn_v1beta1.Deployment{}
+			err = cli.ExtensionsV1beta1().
+				RESTClient().
+				Get().
+				Namespace(namespace).
+				Name(name).
+				Resource("deployments").
+				Do().
+				Into(d)
+			return
+		}
+	}
+
+	if k.getAppsV1 == nil {
+		k.getAppsV1 = func(cli *kubernetes.Clientset, name,
+			namespace string) (d *api_apps_v1.Deployment, err error) {
+			d = &api_apps_v1.Deployment{}
+			err = cli.AppsV1().
+				RESTClient().
+				Get().
+				Namespace(namespace).
+				Name(name).
+				Resource("deployments").
+				Do().
+				Into(d)
+			return
+		}
+	}
+
+	if k.rollOutStatusExtnV1Beta1 == nil {
+		k.rollOutStatusExtnV1Beta1 = func(cli *kubernetes.Clientset, name,
+			namespace string) (op []byte, err error) {
+			d, err := k.getExtnV1Beta1(cli, name, namespace)
+			if err != nil {
+				return nil, err
+			}
+			return DeployExtnV1Beta1(WithExtnV1Beta1Deployment(d)).
+				AddCheck(IsSyncSpecV1B1()).
+				AddCheck(IsProgressDeadlineExceededV1B1()).
+				AddCheck(IsTerminationInProgressV1B1()).
+				AddCheck(IsUpdationInProgressV1B1()).
+				AddCheck(IsOlderReplicaActiveV1B1()).
+				RollOutStatus()
+		}
+	}
+
+	if k.rollOutStatusAppsV1 == nil {
+		k.rollOutStatusAppsV1 = func(cli *kubernetes.Clientset, name,
+			namespace string) (op []byte, err error) {
+			d, err := k.getAppsV1(cli, name, namespace)
+			if err != nil {
+				return nil, err
+			}
+			return DeployAppsv1(WithAppsv1Deployment(d)).
+				AddCheck(IsSyncSpecV1()).
```

Comment:
Let us not add the checks here.
Checks will be registered & used within `func (d *deploy) isRollout() (string, bool)` only
isRollout() will be invoked from within `func (d *deploy) RolloutStatusf() ([]byte, error)`

There is nothing wrong in this approach as builder pattern helps us to achieve this in a neat way.
However, IMO `RolloutStatusf()` is the right place!!!

Reply:
done.